File: python/Recursion&Backtracking/Combination_Sum.py
class Solution:
  def combinationSum(self, candidates, target):
    result = []

    def dfs(index, curr, total):
      if total == target:
        result.append(curr[:])
        return

      if index >= len(candidates) or total > target:
        return
      
      curr.append(candidates[index])
      dfs(index, curr, total + candidates[index])

      curr.pop()
      dfs(index + 1, curr, total)
    
    dfs(0, [], 0)
    return result

File: C++/Recursion&Backtracking/Combination_Sum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  void solve(int idx, int target, vector<int> &candidates, vector<int> &temp, vector<vector<int>> &ans)
  {
    if (target == 0)
    {
      ans.push_back(temp);
      return;
    }

    if (idx == candidates.size() || target < 0)
      return;

    temp.push_back(candidates[idx]);
    solve(idx + 1, target, candidates, temp, ans);
  }

  vector<vector<int>> combinationSum(vector<int> &candidates, int target)
  {
    vector<vector<int>> ans;
    vector<int> temp;

    solve(0, target, candidates, temp, ans);

    return ans;
  }
};